lib/ui/platform_dispatcher.dart

```diff
@@ -2019,17 +2019,23 @@ class Locale {
     if (other is! Locale) {
       return false;
     }
-    final String? countryCode = _countryCode;
+    final String? thisCountryCode = countryCode;
     final String? otherCountryCode = other.countryCode;
     return other.languageCode == languageCode
         && other.scriptCode == scriptCode // scriptCode cannot be ''
-        && (other.countryCode == countryCode // Treat '' as equal to null.
-            || otherCountryCode != null && otherCountryCode.isEmpty && countryCode == null
-            || countryCode != null && countryCode.isEmpty && other.countryCode == null);
+        && (other.countryCode == thisCountryCode // Treat '' as equal to null.
+            || otherCountryCode != null && otherCountryCode.isEmpty && thisCountryCode == null
+            || thisCountryCode != null && thisCountryCode.isEmpty && other.countryCode == null);
   }
 
   @override
-  int get hashCode => hashValues(languageCode, scriptCode, countryCode == '' ? null : countryCode);
+  int get hashCode => _hashCode[this] ??= hashValues(
+        languageCode,
+        scriptCode,
+        countryCode == '' ? null : countryCode,
+      );
+  // Memoize this value since languageCode and countryCode require lookups.
+  static final Expando<int> _hashCode = Expando<int>();
```

Comment:
Expandos aren't free - how much is the lookup in the expando costing vs calculating the hashCode outright?

One thing I wonder about is - locale doesn't change all that often (assumed below in the memoization for toString).

Could we be using a better data structure that would make it cheaper/faster to lookup the last used locale when we do lookups?  I'm not sure where the lookup code is happening though.

Reply:
(I just read your comment in the issue that seems to answer the first part of my question. The second part could be addressed independent from this patch)

Reply:
> Expandos aren't free - how much is the lookup in the expando costing vs calculating the hashCode outright?

Right, napkin math we are exchanging (1 map lookup) for (2 map lookups, 3 calls to hashCode, 1 hashValue calls, 1 null comparison).  I ran a microbenchmark I described in the description and this was 70% faster.  The memory cost is just one int per Locale.  Since Locale is const there shouldn't be many of them.

> One thing I wonder about is - locale doesn't change all that often (assumed below in the memoization for toString).

I don't think it has to do with changing the Locale, looks like the Map of all the Locales is being generated no matter the selected Locale.

> Could we be using a better data structure that would make it cheaper/faster to lookup the last used locale when we do lookups? I'm not sure where the lookup code is happening though.

I'm not sure that would be faster since the comparison of equality to see if you are talking to the last Locale may be more expensive than looking up the Expando which is presumably looking into a map based on an identity hash which is readily available.  The equality check requires 2 map lookups (into the deprecated tables).  Plus it wouldn't help when rebuilding a map of multiple Locales, which seems to be happening in Gallery's case.

Reply:
The way to make it faster would be to check if they're identical, which for const instances that are the same should evaluate to true without doing the operator== check.